Compilation error with MSYS2 mingw-w64, 'M_PI' was not defined in this scope

When compiling pcl on Windows using MSYS2 mimgw-w64 I get a lot of `'M_PI' was not declared in this scope` errors.

Full error log
```
[  0%] Building CXX object common/CMakeFiles/pcl_common.dir/src/bearing_angle_image.cpp.obj
cd /C/msys64/home/filip.sund/code/pcl-fsund/build/common && /C/msys64/mingw64/bin/g++.exe  -DBOOST_LIB_DIAGNOSTIC -DBOOST_THREAD_USE_LIB -DPCLAPI_EXPORTS -DvtkRenderingContext2D_AUTOINIT="1(vtkRenderingContextOpenGL2)" -DvtkRenderingCore_AUTOINIT="3(vtkInteractionStyle,vtkRenderingFreeType,vtkRenderingOpenGL2)" -I/C/msys64/mingw64/include/vtk-8.2 -I/C/msys64/mingw64/include/freetype2 -I/C/msys64/mingw64/include/double-conversion -I/C/msys64/home/filip.sund/code/pcl-fsund/build/include -I/C/msys64/home/filip.sund/code/pcl-fsund/common/include -isystem /C/msys64/mingw64/include/eigen3 -isystem /C/msys64/home/filip.sund/code/pcl-fsund/recognition/include/pcl/recognition/3rdparty  -Wabi=11 -Wall -Wextra -Wno-unknown-pragmas -fno-strict-aliasing -Wno-format-extra-args -Wno-sign-compare -Wno-invalid-offsetof -Wno-conversion -march=native -msse4.2 -mfpmath=sse  -mwin32 -mthreads -O3 -DNDEBUG   -DBOOST_DISABLE_ASSERTS -DEIGEN_NO_DEBUG -std=c++14 -o CMakeFiles/pcl_common.dir/src/bearing_angle_image.cpp.obj -c /C/msys64/home/filip.sund/code/pcl-fsund/common/src/bearing_angle_image.cpp
C:/msys64/home/filip.sund/code/pcl-fsund/common/src/bearing_angle_image.cpp: In member function 'double pcl::BearingAngleImage::getAngle(const pcl::PointXYZ&, const pcl::PointXYZ&)':
C:/msys64/home/filip.sund/code/pcl-fsund/common/src/bearing_angle_image.cpp:78:73: error: 'M_PI' was not declared in this scope
   78 |     theta = std::acos ((a + b - c) / (2 * sqrt (a) * sqrt (b))) * 180 / M_PI;
      |                                                                         ^~~~
make[2]: *** [common/CMakeFiles/pcl_common.dir/build.make:258: common/CMakeFiles/pcl_common.dir/src/bearing_angle_image.cpp.obj] Error 1
make[2]: Leaving directory '/home/filip.sund/code/pcl-fsund/build'
make[1]: *** [CMakeFiles/Makefile2:1033: common/CMakeFiles/pcl_common.dir/all] Error 2
make[1]: Leaving directory '/home/filip.sund/code/pcl-fsund/build'
make: *** [Makefile:152: all] Error 2
```

The first error is in `bearing_angle_image.cpp`, but I see `M_PI` is used in many other places as well.

Exact steps to reproduce are:
 - Install MSYS2 and all dependencies on Windows 10
 - Clone pcl
```
cd pcl
mkdir build
cd build
cmake -G"MSYS Makefiles" -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr ..
make
```

## How to fix
What I did was to just add a top-level `add_definitions(-D_USE_MATH_DEFINES)` to `CMakeLists.txt`, but I realize this is not the proper way to fix this.

I see some trickery is happening on Windows in [pcl_macros.h](https://github.com/PointCloudLibrary/pcl/blob/master/common/include/pcl/pcl_macros.h#L115), but I am not certain if there is a similar way to do it for MSYS2.

~~I think the root issue is that the CMake command `set(CMAKE_CXX_STANDARD 14)` causes the `-std=gnu++14` flag to be added on Linux, but on Windows+MSYS2 the `-std=c++14` flag is added instead. This that `M_PI` is not defined on MSYS2, since that macro is not part of the C++14 standard.~~ This is not true. Since we have `set(CMAKE_CXX_EXTENSIONS OFF)` Linux, the flag is the same on Linux and Windows/MSYS2. 

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* Operating System and version: Windows 10, Version 10.0.18363 Build 18363
* Compiler: g++.exe (Rev2, Built by MSYS2 project) 9.2.0
* PCL Version: commit b1d9c7edd3eddd8f6f8626919b4667869459ceb7

#3545 did not fix the issue, compilation fails in the same place as before

```
Building CXX object common/CMakeFiles/pcl_common.dir/src/bearing_angle_image.cpp.obj
cd /C/msys64/home/filip.sund/code/pcl/build/common && /C/msys64/mingw64/bin/g++.exe  -DBOOST_LIB_DIAGNOSTIC -DBOOST_THREAD_USE_LIB -DPCLAPI_EXPORTS -DPCL_ONLY_CORE_POINT_TYPES -DvtkRenderingContext2D_AUTOINIT="1(vtkRenderingContextOpenGL2)" -DvtkRenderingCore_AUTOINIT="3(vtkInteractionStyle,vtkRenderingFreeType,vtkRenderingOpenGL2)" -I/C/msys64/mingw64/include/vtk-8.2 -I/C/msys64/mingw64/include/freetype2 -I/C/msys64/mingw64/include/double-conversion -I/C/msys64/home/filip.sund/code/pcl/build/include -I/C/msys64/home/filip.sund/code/pcl/common/include -isystem /C/msys64/mingw64/include/eigen3 -isystem /C/msys64/home/filip.sund/code/pcl/recognition/include/pcl/recognition/3rdparty  -Wabi=11 -Wall -Wextra -Wno-unknown-pragmas -fno-strict-aliasing -Wno-format-extra-args -Wno-sign-compare -Wno-invalid-offsetof -Wno-conversion -march=native -msse4.2 -mfpmath=sse  -mwin32 -mthreads -O3 -DNDEBUG   -DBOOST_DISABLE_ASSERTS -DEIGEN_NO_DEBUG -std=c++14 -o CMakeFiles/pcl_common.dir/src/bearing_angle_image.cpp.obj -c /C/msys64/home/filip.sund/code/pcl/common/src/bearing_angle_image.cpp
C:/msys64/home/filip.sund/code/pcl/common/src/bearing_angle_image.cpp: In member function 'double pcl::BearingAngleImage::getAngle(const pcl::PointXYZ&, const pcl::PointXYZ&)':
C:/msys64/home/filip.sund/code/pcl/common/src/bearing_angle_image.cpp:84:73: error: 'M_PI' was not declared in this scope
   84 |     theta = std::acos ((a + b - c) / (2 * sqrt (a) * sqrt (b))) * 180 / M_PI;
      |                                                                         ^~~~
```

--- a/common/include/pcl/pcl_macros.h
+++ b/common/include/pcl/pcl_macros.h
@@ -56,7 +56,7 @@
 #endif
 
 #ifndef _USE_MATH_DEFINES
-#define _USE_MATH_DEFINES
+  #define _USE_MATH_DEFINES
 #endif
 #include <cmath>
 #include <cstdarg>
@@ -118,11 +118,13 @@ namespace pcl
   using int_fast16_t PCL_DEPRECATED("use std::int_fast16_t instead of pcl::int_fast16_t") = std::int_fast16_t;
 }
 
-#if defined _WIN32 && defined _MSC_VER
+#if defined _WIN32
 
 // Define math constants, without including math.h, to prevent polluting global namespace with old math methods
 // Copied from math.h
-#ifndef _MATH_DEFINES_DEFINED
+// Check for M_2_SQRTPI since the cmath header on mingw-w64 doesn't seem to define
+// _MATH_DEFINES_DEFINED when included with _USE_MATH_DEFINES
+#if !defined _MATH_DEFINES_DEFINED && !defined M_2_SQRTPI
   #define _MATH_DEFINES_DEFINED
 
   #define M_E        2.71828182845904523536   // e
@@ -140,15 +142,16 @@ namespace pcl
   #define M_SQRT1_2  0.707106781186547524401  // 1/sqrt(2)
 #endif
 
-// Stupid. This should be removed when all the PCL dependencies have min/max fixed.
-#ifndef NOMINMAX
-# define NOMINMAX
+#if defined _MSC_VER
+  // Stupid. This should be removed when all the PCL dependencies have min/max fixed.
+  #ifndef NOMINMAX
+    #define NOMINMAX
+  #endif
+
+  #define __PRETTY_FUNCTION__ __FUNCTION__
+  #define __func__ __FUNCTION__
 #endif
-
-# define __PRETTY_FUNCTION__ __FUNCTION__
-# define __func__ __FUNCTION__
-
-#endif //defined _WIN32 && defined _MSC_VER
+#endif // defined _WIN32
 
 
 template<typename T>

